Isolate macOS wheel builds from Homebrew

freakboy3742 opened the pull request:
The existing macOS `cibuildwheel` configuration relies on Homebrew to provide *some* dependencies. This clearly works in practice, but there are some issues associated with this choice.

Firstly, it is destructive when used on a local build machine. The wheel dependency script invokes `brew remove --ignore-dependencies` (which can leave the host machine in a broken, and potentially difficult to restore state); and requires the use of `sudo` to make changes in the `/usr/local` tree. This means it is difficult to test `cibuildwheel` changes locally, or to recommend using `cibuildwheel` as a local build solution.

Secondly, Homebrew builds could be incompatible with Pillow builds. The `MACOSX_DEPLOYMENT_TARGET` for Homebrew dependencies is fixed by the Homebrew build chain, rather than the Pillow build tools, so a change in Pillow's configuration may not be satisfied by Homebrew's build configuration. This isn't an issue at present, but it could easily become one in future.

Related to this, there is also the potential that a change in a Homebrew recipe could lead to the inadvertent introduction of additional binary libraries into the delocated macOS binaries. This is especially problematic with the fribidi, raqm and freetype dependedencies which are included as "vendor" sources.

However, the *real* motivation for proposing this change is that it is a precursor for PEP 730-compliant iOS builds. If Homebrew is on the build path when compiling iOS binaries, compiler tooling will often find an ARM64 Homebrew binary and attempt link it into an iOS library - which then (predictably) breaks. So - Homebrew isolation is essential for iOS compilation. Since the build processes for iOS and macOS are very similar, and Homebrew isolation was necessary for iOS, adding Homebrew isolation for macOS will simplify any future iOS patch. It also provides a way to submit iOS support in a series of smaller pieces, rather than a single "monster" patch. Plus, it provides all the side benefits of isolation described above.

An overview of the changes:

* Uses the `build` folder as a location for all dependency builds. This prevents the root checkout from becoming dirtied by dependency artefacts, simplifying cleanup, and avoiding the need for extensive modifications to the `.gitignore` file.

* Sets up an isolated build prefix in the `build` folder (`./build/deps/darwin`) where dependencies can be installed, and `make installs` all dependencies into that path.

* Forces `PATH` to be a "clean" environment that only includes bare system tools, plus the Python binary being used for the build, and the isolated build prefix. macOS *doesn't* include most of it's development libraries in /usr or /usr/local, instead using a path provided by the macOS SDK (which is configured as part of the compiler toolchain).

* Adds a build of `pkg-config` to the dependencies. This is the one build tool that autotools and cmake often require that Xcode doesn't provide. Providing a custom build of `pkg-config` also ensures that no dependencies other than the ones provided by cibuildwheel are included. The build recipe that is used is derived from the Homebrew recipe.

* ~Adds a build of `libXdmcp`, to avoid a dependency on the Homebrew-provided version.~

* ~Adds builds of `fribidi` and `raqm`, rather than using the Homebrew versions as "vendored" versions. The `setup.py` configuration passed in by `cibuildwheel` has been modified from the default on other platforms to *not* use the `vendor` version.~

* Updates the pinned multibuild version to the current `devel` branch hash. This is to incorporate Python3.13 support, plus a number of fixes required to support installs into locations other than `/usr/local`, and corrections to LIBDIR handling for platforms that use the `lib64` suffix.

* ~Adds a `dependencies-prefix` configuration option to `setup.py`. This allows the user to provide a specific location to look for dependencies, rather than the series of Fink/MacPorts/Homebrew fallbacks that currently exist.~

radarhere commented on .github/workflows/wheels-test.sh:
```diff
@@ -1,16 +1,13 @@
-    export PKG_CONFIG_PATH="/usr/local/opt/openblas/lib/pkgconfig"
-    if [ -f /opt/homebrew/lib/libfribidi.dylib ]; then
-        sudo cp /opt/homebrew/lib/libfribidi.dylib /usr/local/lib
+# For Unix, Ensure fibidi is installed by the system.
```
```suggestion
# For Unix, ensure fribidi is installed by the system.
```

radarhere commented on .github/workflows/wheels-dependencies.sh:
```diff
@@ -23,6 +23,8 @@ OPENJPEG_VERSION=2.5.2
 XZ_VERSION=5.6.3
 TIFF_VERSION=4.6.0
 LCMS2_VERSION=2.16
+RAQM_VERSION=0.7.1
```
```suggestion
RAQM_VERSION=0.10.2
```

freakboy3742 replied:
It definitely makes sense to use the most recent available version; but pillow-depends-main doesn't include this version. What's the procedure for updating that repo?

(Related: pkg-config, libXdmcp, and an updated webp sources should probably be added to that project)

radarhere replied:
The theory is that pillow-depends-main is just a more reliable source for the dependencies, and PRs should work without it.

Once this is merged, I will likely add them there.

freakboy3742 commented:
The manylinux failures look like they'll be fixed by [this PR](https://github.com/multi-build/multibuild/pull/537) to multi build.

The x86_64 macOS failures are probably a leakage from /usr/local that I haven't accounted for. I'm looking into that one now.

I'm not sure how to interpret the coverage drop, though...

freakboy3742 commented:
@radarhere I've updated the pin for multibuild to include the (just merged) PR reverting the libjpeg-turbo CMAKE_INSTALL_LIBDIR change; and I've addressed some additional locations where Homebrew can leak in on x86_64.

nulano commented:
> Related to this, there is also the potential that a change in a Homebrew recipe could lead to the inadvertent introduction of additional binary libraries into the delocated macOS binaries. 

This has already happened many times in the past, so thanks for working on this.

> Adds builds of fribidi and raqm, rather than using the Homebrew versions as "vendored" versions. The setup.py configuration passed in by cibuildwheel has been modified from the default on other platforms to not use the vendor version.

I'm not sure if I understand what you mean here. We need to use a vendored version of raqm in wheels for license reasons. When using `raqm=vendor`, we build our own version of raqm from https://github.com/python-pillow/Pillow/tree/main/src/thirdparty/raqm. Similarly, using `fribidi=vendor` does not actually build fribidi, but instead custom code that can load fribidi at runtime, see https://github.com/python-pillow/Pillow/tree/main/src/thirdparty/fribidi-shim. The Homebrew version of fribidi is only used for testing the built wheels.

freakboy3742 commented:
> > Adds builds of fribidi and raqm, rather than using the Homebrew versions as "vendored" versions. The setup.py configuration passed in by cibuildwheel has been modified from the default on other platforms to not use the vendor version.
> 
> I'm not sure if I understand what you mean here. We need to use a vendored version of raqm in wheels for license reasons. When using `raqm=vendor`, we build our own version of raqm from https://github.com/python-pillow/Pillow/tree/main/src/thirdparty/raqm. Similarly, using `fribidi=vendor` does not actually build fribidi, but instead custom code that can load fribidi at runtime, see https://github.com/python-pillow/Pillow/tree/main/src/thirdparty/fribidi-shim. The Homebrew version of fribidi is only used for testing the built wheels.

Thanks - that's some helpful context. I was seeing the missing fribidi detection during the build and interpreting that as something that was part of the move away from homebrew; but I see now that Pillow is loading from Homebrew *by design* in this instance. That also might explain the coverage drop (as supplying fribidi means any runtime loading code won't be in use). I'll take another swing at that part (and resolve the linux build issues as well).

Out of interest - what's the license issue with raqm? The source code indicates It looks to be MIT licensed... did it used to be GPL (or something else problematic?)

hugovk commented:
> Out of interest - what's the license issue with raqm? The source code indicates It looks to be MIT licensed... did it used to be GPL (or something else problematic?)

From https://github.com/python-pillow/Pillow/pull/2753 in 2017:

> We can't currently distribute a binary of Pillow with support for raqm due to license incompatibility, because while libraqm is MIT licensed, it links to GPL libraries.

FriBiDi is LGPL. 

Since then, Raqm added support for Apache-licensed SheenBiDi in 2021: https://github.com/HOST-Oman/libraqm/issues/138, https://github.com/HOST-Oman/libraqm/pull/139.

Perhaps we could switch.

freakboy3742 commented:
@radarhere After a lot of false starts (apologies for the noise, BTW...), I think I've finally got it all sorted. 

1. macOS is isolated from Homebrew on both x86_64 and ARM64
2. The shim fribidi builds is used in the wheel, with Homebrew providing the fribidi binary on macOS in the test environment
3. Linux builds are now correctly using `BUILD_PREFIX`, and don't require the `lib64` copying step.

radarhere commented on .github/workflows/wheels-test.sh:
```diff
@@ -1,12 +1,20 @@
-    if [ -f /opt/homebrew/lib/libfribidi.dylib ]; then
-        sudo cp /opt/homebrew/lib/libfribidi.dylib /usr/local/lib
+    # If Homebrew is on the path during the build, it may leak into the wheels.
+    # However, we need a *do* need Homebrew to provide a copy of fribidi for
```
```suggestion
    # However, we *do* need Homebrew to provide a copy of fribidi for
```

freakboy3742 commented:
I'm not sure I understand how coverage can go *down* when you don't touch any code or tests... and the codecov report isn't really helping explain what is going on.

hugovk commented:
Hmm, let's see... (Short version: I think it's some problem with Codecov and not this PR.)

<details>
<summary>Codecov details</summary>

Looking at the coverage report at https://github.com/python-pillow/Pillow/pull/8497/checks?check_run_id=32242849880 it says "90.36% (-3.98%) compared to 2d1d801" which is already suspicious: our coverage is around 90%, not 94%.

That 2d1d801 is the last commit on `main`: "Update CHANGES.rst [ci skip]"

That "[ci skip]" did in fact skip the CI, but I see @radarhere later triggered a manual build of that commit at https://github.com/python-pillow/Pillow/actions/runs/11573601676 for only Windows, presumably checking the PyPy failure was also happening on `main` and not just on (another) failing PR.

Anyway, this Windows-only CI resulted in a skewed 94% coverage report at https://app.codecov.io/github/python-pillow/Pillow/commit/2d1d801ec032bfb760bede88da6dcaab5ff3c75e (see the flags on the right only show Windows and not the others).

All of which suggests this is nothing to worry about. Having said that, these jobs ran about 9 hours ago and @freakboy3742's comment is about 14 hours ago, so this digging doesn't reflect the situation 14 hours ago!

I will say that we've been seeing drops in other PRs where it doesn't quite make sense. I've not had time to investigate, I wonder if we're not setting tokens properly for some builds after Codecov changed their v4 action to be more strict.

</details>

freakboy3742 commented:
@hugovk /me casually drops [this article](https://hynek.me/articles/ditch-codecov-python/) into the chat :-)

FWIW - BeeWare ditched Codecov a couple of years ago specifically because of weird inconsistencies like this one; and although the output of a raw coverage report is marginally less pretty (i.e., no annotated code listings), you can get the same report locally that you get in CI, and you get *none* of the weird outages and reporting issues that seem to plague Coverage. YMMV; but if there's interest, it's a relatively straightforward switch, especially if coverage reporting is already in place.

freakboy3742 commented:
The (seemingly inconsistent) Windows/pypy3.10 test failure also fascinates me... is this the same cache issue mentioned by #8513?

radarhere commented on .github/workflows/wheels-test.sh:
```diff
@@ -1,12 +1,20 @@
+    # However, we *do* need Homebrew to provide a copy of fribidi for
+    # testing purposes so that we can verify the fribidi shim works as expected.
+    if [[ "$(uname -m)" == "x86_64" ]]; then
+        HOMEBREW_HOME=/usr/local/homebrew
```
```suggestion
        HOMEBREW_HOME=/usr/local
```
I would have thought just '/usr/local', based on https://docs.brew.sh/Installation
> The script installs Homebrew to its default, supported, best prefix (/opt/homebrew for Apple Silicon, /usr/local for macOS Intel

freakboy3742 replied:
Turns out both work (for `brew`) at least - Homebrew keeps the "original" versions in /usr/local/homebrew. However, I agree that `/usr/local` is the generally intended entry point.

freakboy3742 replied:
I take that back - it *does* matter (and making this change caused test failures on x86_64).

Firstly, it's possible for homebrew to have packages *installed*, but not *linked*. Using the `/usr/lib/Homebrew` location ensures that the binary can be used regardless (note also the capitalisation fix - doesn't strictly matter on HFS+, but for accuracy).

Secondly, it's possible for *other* libraries to be in `/usr/local/lib`. Using `/usr/local/Homebrew/lib` for the `DYLD_LIBRARY_PATH` avoids this.

hugovk commented:
> The (seemingly inconsistent) Windows/pypy3.10 test failure also fascinates me... is this the same cache issue mentioned by #8513?

Yep, we added a new licence (well, old: MIT CMU) to OSI and Trove classifiers, so that test needs the new `trove-classifiers`. https://github.com/python-pillow/Pillow/issues/7942 has the details.

I've opened https://github.com/python-pillow/Pillow/pull/8514 as an alternative to https://github.com/python-pillow/Pillow/pull/8513.

nulano commented:
> * The shim fribidi builds is used in the wheel, with Homebrew providing the fribidi binary on macOS in the test environment

Thanks, that looks better.

---
There still seems to be something wrong with freetype on arm64, https://github.com/python-pillow/Pillow/actions/runs/11604707482/job/32314276332?pr=8497#step:5:14308
```
  SKIPPED [2] Tests/test_imagefont.py:1081: FreeType compiled without brotli or WOFF2 support
```
But it was built with brotli, https://github.com/python-pillow/Pillow/actions/runs/11604707482/job/32314276332?pr=8497#step:5:7768:
```
    brotli:        yes (pkg-config)
```
We need to support this for #6554.
---
We could also consider compiling libwebp before libtiff so that WEBP-compressed tiff files can be read (AFAIK webp only looks for libtiff so that it can compile companion tools, but does not actually use it in the library - as I mentioned in #6562). But we don't need to do it now since the test was already being skipped on macOS and Linux before this PR.
https://github.com/python-pillow/Pillow/actions/runs/11604707482/job/32314276332?pr=8497#step:5:15489:
```
  SKIPPED [1] Tests/test_file_libtiff.py:905: WEBP compression support is not configured
```

freakboy3742 commented:
> There still seems to be something wrong with freetype on arm64, https://github.com/python-pillow/Pillow/actions/runs/11604707482/job/32314276332?pr=8497#step:5:14308

Weird. My immediate guess is that it might have something to do with the fact that a vendored version of fribidi is being used, which might *not* be triggering brotli support (or at least not triggering the test). I'll take a look and see what is going on. 

> We could also consider compiling libwebp before libtiff so that WEBP-compressed tiff files can be read (AFAIK webp only looks for libtiff so that it can compile companion tools, but does not actually use it in the library - as I mentioned in #6562). But we don't need to do it now since the test was already being skipped on macOS and Linux before this PR. https://github.com/python-pillow/Pillow/actions/runs/11604707482/job/32314276332?pr=8497#step:5:15489:

I saw that, and ultimately gave up trying to fix it - firstly because I was trying to limit the scope of the changes, but also because I'm not sure it *can* be fixed. libwebp has a dependency on tiff... but then lcms2 needs tiff; and openjpeg needs libpng, tiff, and lcms2. So - I'm not sure this *can* be fixed (or, at least, it can't be fixed without some sort of 2-pass compile and the use of dynamic linking, which iOS can't use anyway).

nulano commented:
> Weird. My immediate guess is that it might have something to do with the fact that a vendored version of fribidi is being used, which might not be triggering brotli support (or at least not triggering the test). I'll take a look and see what is going on.

The test skip is triggered by a specific error raised by the freetype library which itself does not use fribidi.

> libwebp has a dependency on tiff... but then lcms2 needs tiff; and openjpeg needs libpng, tiff, and lcms2

AFAIK all of these dependencies are just for the binary utilities that we don't need so we can disable them with relevant compile flags; the libraries themselves do not have dependencies between them (except for libtiff optionally using libwebp). But I agree that it might be best to leave that for a separate PR.

freakboy3742 commented:
@nulano I've found the culprit - Homebrew's copy of freetype was leaking into the test environment.

By using `DYLD_LIBRARY_PATH` pointed at Homebrew's lib foldler, *all* of Homebrew's libraries become available, and take take precedence over the ones provided by the wheel. As a result, Homebrew's copy of freetype is used, which *doesn't* include Brotli support. I presume this wasn't an issue previously because an explicit `DYLD_LIBRARY_PATH` comes earlier in library resolution precedence; when Homebrew was being used implicitly, Homebrew's library resolve *after* Pillow's bundled dependencies.

This can be fixed by pointing the `DYLD_LIBRARY_PATH` at the *cellar* version of the vendored fribidi library (which is the only library we actually need from Homebrew). I've pushed that update (update: I'm looking at the linting error now...)

The update includes 2 other small changes:

1. The location of the "homebrew isolated" macOS build is slightly altered - the extra `darwin` folder became necessary when I started working on the iOS build, and the build folders for macOS started colliding with iOS. 
2. I've added some safety checks against some uses that are valid cibuildwheel invocations, but will break if you try to use them. Specifically, `CIBW_ARCHS="arm64 x86_64"`, and not specifying `CIBW_ARCHS` at all, are both valid cibuildwheel setup, but Pillow doesn't allow it because the `before_all` step requires a single explicit architecture. This bit me a couple of times before I worked out what was happening; I figured it couldn't hurt to raise an explicit error when we know it won't work.

radarhere commented on .github/workflows/wheels-dependencies.sh:
```diff
@@ -1,11 +1,34 @@
+PROJECTDIR=$(pwd)
+if [[ "$(uname -s)" == "Darwin" ]]; then
+    # Safety check - macOS builds require that CIBW_ARCHS is set, and that it
+    # only contains a single value (even though cwbuildwheel) allows multiple
```
```suggestion
    # only contains a single value (even though cibuildwheel) allows multiple
```

nulano commented:
Thanks, looks good to me now.

radarhere commented on .github/workflows/wheels-dependencies.sh:
```diff
@@ -1,11 +1,34 @@
+PROJECTDIR=$(pwd)
+if [[ "$(uname -s)" == "Darwin" ]]; then
+    # Safety check - macOS builds require that CIBW_ARCHS is set, and that it
+    # only contains a single value (even though cibuildwheel) allows multiple
+    # values in CIBW_ARCHS.
```
```suggestion
    # only contains a single value (even though cibuildwheel allows multiple
    # values in CIBW_ARCHS).
```

radarhere commented on .github/workflows/wheels-dependencies.sh:
```diff
@@ -77,67 +109,88 @@ function build {
+    # with a MACOSX_DEPLOYMENT_TARGET that doesn't match what we want to use,
+    # and they may bring in other dependencies that we don't want. The same will
+    # be true of any other locations on the path. To avoid conflicts, strip the
+    # path down to the bare mimimum (which, on macOS, won't include any
```
```suggestion
    # path down to the bare minimum (which, on macOS, won't include any
```

radarhere commented on setup.py:
```diff
@@ -448,7 +454,7 @@ def _remove_extension(self, name: str) -> None:
         try:
             sdk_path = (
-                subprocess.check_output(["xcrun", "--show-sdk-path"])
+                subprocess.check_output(["xcrun", "--show-sdk-path", "--sdk", "macosx"])
```
What was the reason for this change?

freakboy3742 replied:
Explicitness. Xcode can manage multiple SDKs (iOS being an obvious one); adding the explicit `--sdk macosx` SDK ensures that you're *definitely* getting the macOS one. That will be the default on *almost* every install of Xcode, but if you have a stray `SDKROOT` environment variable from some other activity, it could be inadvertently pointing at an iOS, tvOS, visionOS or MacCatalyst SDK.

radarhere commented on .github/workflows/wheels-test.sh:
```diff
@@ -1,12 +1,27 @@
+    # installed copy of fribidi is cellared. This ensures we don't pick up the
+    # Homebrew version of any other library that we're dependent on (most notably,
+    # freetype).
+    export DYLD_LIBRARY_PATH=$(dirname $(realpath $HOMEBREW_HOME/lib/libfribidi.dylib))
```
```suggestion
    if [[ "$(uname -m)" == "x86_64" ]]; then
        FRIBIDI_DYLIB=$(find /usr/local/Cellar/fribidi -type f -name libfribidi.dylib)
    else
        FRIBIDI_DYLIB=$HOMEBREW_HOME/lib/libfribidi.dylib
    fi
    export DYLD_LIBRARY_PATH=$(dirname $FRIBIDI_DYLIB)
```

https://github.com/python-pillow/Pillow/actions/runs/11752543578/job/32744256384?pr=8497#step:5:9142
> realpath: /usr/local/Homebrew/lib/libfribidi.dylib: No such file or directory

libfribidi.dylib isn't located there on the macOS Intel jobs.

freakboy3742 replied:
Huh... at some point, I've clearly done something to mess up my x86 test machine's Homebrew install, because I've got a `/usr/local/Homebrew/Cellar`... but I now see I'm also getting a bunch of warnings because of the existence of that folder. 

The `find`-based solution you've got here will fail (or be prone to failure) if there's more than one version of fribidi installed, which is possible on user machines; but if `HOMEBREW_HOME` is moved back to `/usr/local`, the `realpath` approach should work.

radarhere commented on .github/workflows/wheels-dependencies.sh:
```diff
@@ -38,35 +60,44 @@ BZIP2_VERSION=1.0.8
 
+function build_pkg_config {
+    if [ -e pkg-config-stamp ]; then return; fi
+    # This essentially duplicates the Homebrew recipe:
+    # https://github.com/Homebrew/homebrew-core/blob/master/Formula/p/pkg-config.rb
```
```suggestion
    # This essentially duplicates the Homebrew recipe
```
https://github.com/Homebrew/homebrew-core/blob/master/Formula/p/pkg-config.rb no longer exists after https://github.com/Homebrew/homebrew-core/pull/194885

freakboy3742 replied:
Link rot before the PR even lands has to be a new record... :-)

I'll remove the reference.

radarhere commented on .github/workflows/wheels-dependencies.sh:
```diff
@@ -77,67 +108,88 @@ function build {
     CFLAGS="$CFLAGS -O3 -DNDEBUG"
-    if [[ -n "$IS_MACOS" ]]; then
-        CFLAGS="$CFLAGS -Wl,-headerpad_max_install_names"
-    fi
```
What was your thinking behind removing this? I'm just concerned that https://github.com/python-pillow/Pillow/issues/7250 might re-appear.

freakboy3742 replied:
This got removed when I was getting webp to build, because it didn't seem to causing any build issues. I see now it's related to py2app packaging, so wouldn't necessarily manifest in a CI failure. I'll restore it.

radarhere commented on .github/workflows/wheels-dependencies.sh:
```diff
@@ -77,67 +108,88 @@ function build {
+    export PATH="$BUILD_PREFIX/bin:$(dirname $(which python3)):/usr/bin:/bin:/usr/sbin:/sbin:/Library/Apple/usr/bin"
+    export CMAKE_PREFIX_PATH=$BUILD_PREFIX
+
+    # Link the brew command into our isolated build directory.
```
I'm not following this comment. Are you talking about a specific brew command that we're invoking, or just if a user wants to add their own call to brew?

freakboy3742 replied:
A stale comment. In an earlier iteration, it was explicitly linking the brew executable into `$BUILD_PREFIX/bin`; that's no longer happening. I'll update the comment.

radarhere commented on .github/workflows/wheels-dependencies.sh:
```diff
@@ -77,67 +108,88 @@ function build {
     if [ -n "$IS_MACOS" ]; then
         # Custom freetype build
-        build_simple freetype $FREETYPE_VERSION https://download.savannah.gnu.org/releases/freetype tar.gz --with-harfbuzz=no
+        build_simple freetype $FREETYPE_VERSION https://download.savannah.gnu.org/releases/freetype tar.gz --without-harfbuzz
```
https://gitlab.freedesktop.org/freetype/freetype/-/blob/master/builds/unix/configure.raw?ref_type=heads#L429 mentions `--with-harfbuzz`, but not `--without-harfbuzz`. What was the thinking here?

freakboy3742 replied:
"--with/--without" handing is a standard feature of autoconf, AFAIK; I think this change got made when I was trying to diagnose some harfbuzz issues. Happy to revert, as it's code churn with no explicit purpose.

radarhere commented on .github/workflows/wheels-dependencies.sh:
```diff
@@ -77,67 +108,88 @@ function build {
     if [ -n "$IS_MACOS" ]; then
         build_simple xorgproto 2024.1 https://www.x.org/pub/individual/proto
         build_simple libXau 1.0.11 https://www.x.org/pub/individual/lib
+        build_simple libXdmcp 1.1.5 https://www.x.org/pub/individual/lib
```
> Adds a build of `libXdmcp`, to avoid a dependency on the Homebrew-provided version.

I suspect you took out all the `brew remove` statements, then noticed this was being pulled from brew, and then added it here to replace that.

I'm saying that this wouldn't be part of the Pillow wheels at the moment, so if it's still not after this PR, that sounds fine to me.

This function
https://github.com/python-pillow/Pillow/blob/5bff2f3b2894ec6923c590d0c37b18177d0634bd/src/display.c#L834 is the only use of XCB in Pillow, so if it's not necessary, I don't think it needs to be here?

freakboy3742 replied:
My recollection was that Xdmcp was needed for XCB to compile *at all*; but I've just removed it and done a clean build, and everything worked as expected (AFAICT). I'll remove it.

radarhere commented on .github/workflows/wheels-dependencies.sh:
```diff
@@ -77,67 +108,88 @@ function build {
+        # Homebrew versions of those libraries from /usr/local.
+        build_simple tiff $TIFF_VERSION https://download.osgeo.org/libtiff tar.gz \
+            --with-jpeg-include-dir=$BUILD_PREFIX/include --with-jpeg-lib-dir=$BUILD_PREFIX/lib \
+            --disable-webp --disable-zstd --disable-libdeflate
```
```suggestion
            --disable-webp --disable-libdeflate --disable-zstd
```
Nitpick - use the arguments in the same order as they are mentioned in the comment.

radarhere commented:
Thanks for your patience. This is getting close - aside from my latest comments, I'm happy with the current state.

freakboy3742 commented:
@radarhere No problem at all - thanks for all the reviews. I've just pushed an update that I think addresses all your comments.

aclark4life commented:
Thanks @freakboy3742 @radarhere @nulano !

hugovk commented:
Thank you!
